bench_runner/scripts/workflow.py

```diff
@@ -0,0 +1,376 @@
+            depth=1,
+        )
+
+
+def compile_unix(cpython: PathLike, flags: list[str], pgo: bool, pystats: bool) -> None:
+    cpython = Path(cpython)
+
+    env = os.environ.copy()
+    if "CLANG" in flags:
+        if sys.platform.startswith("linux"):
+            env["CC"] = util.safe_which("clang-19")
+            env["LLVM_AR"] = util.safe_which("llvm-ar-19")
+            env["LLVM_PROFDATA"] = util.safe_which("llvm-profdata-19")
+        elif sys.platform == "darwin":
+            llvm_prefix = util.get_brew_prefix("llvm")
+            env["CC"] = f"{llvm_prefix}/bin/clang"
+            env["LDFLAGS"] = f"-L{llvm_prefix}/lib"
+            env["CFLAGS"] = f"-I{llvm_prefix}/include"
+
+    if sys.platform == "darwin":
+        openssl_prefix = util.get_brew_prefix("openssl@1.1")
+        env["PKG_CONFIG_PATH"] = f"{openssl_prefix}/lib/pkgconfig"
+
+    args = []
+    if pystats:
+        args.append("--with-pystats")
+    if pgo:
+        args.extend(["--enable-optimizations", "--with-lto=full"])
+    if "PYTHON_UOPS" in flags:
+        assert "JIT" not in flags
+        args.append("--enable-experimental-jit=interpreter")
+    if "JIT" in flags:
+        assert "PYTHON_UOPS" not in flags
+        args.append("--enable-experimental-jit=yes")
+    if "NOGIL" in flags:
+        args.append("--disable-gil")
+    if "CLANG" in flags:
+        args.append("--with-tail-call-interp")
+
+    with contextlib.chdir(cpython):
+        subprocess.check_call(["./configure", *args], env=env)
+        subprocess.check_call(["make", "-j"], env=env)
+
+
+def compile_windows(
+    cpython: PathLike, flags: list[str], pgo: bool, force_32bit: bool
+) -> None:
+    cpython = Path(cpython)
+
+    args = []
+    if force_32bit:
+        args.extend(["-p", "win32"])
+    args.extend(["-c", "Release"])
+    if pgo:
+        args.append("--pgo")
+    if "JIT" in flags:
+        args.append("--experimental-jit")
+    if "PYTHON_UOPS" in flags:
+        args.append("--experimental-jit-interpreter")
+    if "NOGIL" in flags:
+        args.append("--disable-gil")
+    if "CLANG" in flags:
+        args.extend(
+            [
+                "--tail-call-interp",
+                r"/p:PlatformToolset=clangcl",
+                r"/p:LLVMInstallDir=C:\Program Files\LLVM",
+                r"/p:LLVMToolsVersion=19.1.6",
+            ]
+        )
+
+    with contextlib.chdir(cpython):
+        subprocess.check_call(
+            [
+                Path("PCbuild") / "build.bat",
+                *args,
+            ]
+        )
+        shutil.copytree(get_windows_build_dir(force_32bit), "libs", dirs_exist_ok=True)
+
+
+def install_pyperformance(venv: PathLike) -> None:
+    run_in_venv(venv, "pip", ["install", "./pyperformance"])
+
+
+def tune_system(venv: PathLike, perf: bool) -> None:
+    # System tuning is Linux only
+    if not sys.platform.startswith("linux"):
+        return
+
+    run_in_venv(venv, "pyperf", ["system", perf and "reset" or "tune"], sudo=True)
```

Comment:
Does this mean that the CPU frequency scaling governors are set to "powersave" when profiling and "performance" when benchmarking?
Could this not lead to different characteristics between the profiled run and the scored run? Though I admit that the profiling will create different characteristics anyway. So I guess my question is just, what is the reason for this?

Reply:
The reason for this is pretty recent.  When @mpage added events to pyperf so that we can turn perf on and off tightly around benchmarking code, I discovered that the tuning made some of these events not fire all the time, probably due to a lower sampling rate.  This resolved that and did not seem to have a significant effect on the overall perf results.

But you're right in general that the perf measurement makes the results slightly different anyway, which is unavoidable.